### DISCUSSION TITLE
Frictional contact with rigid body and fem

### DISCUSSION THREAD

**Discussion by Erol-hub** [Computational-Mechanics-Material-Models/chrono-mechanics#8] [Q&A] [answered]
Hi, i am trying to define a frictional contact between rigid body (chbody) and finite element nodes. I have two rigid plates one is bottom the other one is top plate. I have a cube shaped fem mesh created by using Euler Beam elements.  I define fricional contact using AddFacesFromBoundary:

auto mcontactsurf = chrono_types::make_shared<ChContactSurfaceMesh>(mysurfmaterial);
my_mesh->AddContactSurface(mcontactsurf);
mcontactsurf->AddFacesFromBoundary(0.0);

I applied prescribed displatement on the top plate using ChLinkMotionImposed class:

auto f_xyz = chrono_types::make_shared<ChFunctionPosition_XYZfunctions>();    
f_xyz->SetFunctionY(chrono_types::make_shared<ChFunction_Ramp>(0,-10.));
auto impose_1 = chrono_types::make_shared<ChLinkMotionImposed>();
sys.Add(impose_1);
    
impose_1->Initialize(top_plate, mtruss, ChFrame<>((1,0,0),0));
impose_1->SetPositionFunction(f_xyz);

But system diverges immediately.  Thanks,

**Comment by mtroemner** [comment, marked as answer]
From Alessandro Tasora:

"To mix FEA and contacts:

- if you use ChSystemSMC:
   - the stiffness of the material of contact plays a relevant role on the divergence of the solver, you may want to make contacts softer and softer until you have a convergent simulation
   - look demo_FEA_contacts_SMC.cpp   for an example
- if you use ChSystemNSC:

   - the only solver that is capable of mixing FEA and non-smooth contacts is ChSolverADMM

   - the ChSolverADMM takes a big benefit if using MKL PARDISO for internal loops, but you must enable the MKL PARDISO module in CMake.

   - for example:

    auto solver = chrono_types::make_shared<ChSolverADMM>(chrono_types::make_shared<ChSolverPardisoMKL>());
    solver->EnableWarmStart(true);
    solver->SetMaxIterations(60);
    solver->SetToleranceDual(1e-4);
    solver->SetTolerancePrimal(1e-4);
    solver->SetRho(1);
    solver->SetStepAdjustPolicy(ChSolverADMM::AdmmStepType::BALANCED_UNSCALED);
    sys.SetSolver(solver);

This said, if your simulation involves more than 10k particles, probably you should better use the new DEM-E module of Chrono (that, however, has some limitations in that it does not support FEA, so if you need also finite elements + particles, with DEM-E you can do it only via cosimulation - one process runs particles, the other runs FEA, then you share forces/positions in synchronization steps... but this is quite complex)"
